core/tests/webdriverio_desktop/coreEditorAndPlayerFeatures.js

```diff
@@ -411,8 +411,50 @@ describe('Core exploration functionality', function() {
       await explorationPlayerPage.expectExplorationToBeOver();
     });
 
+  it('should not show self-loop warning when current and ' +
+    'destination states are different', async function() {
+    await explorationEditorMainTab.setContent(
+      await forms.toRichText('some content'));
+    await explorationEditorMainTab.setInteraction('NumericInput');
+    await explorationEditorMainTab.addResponse(
+      'NumericInput', async function(richTextEditor) {
+        await richTextEditor.appendBoldText('correct');
+      }, 'final card', true, 'IsInclusivelyBetween', -1, 3);
+    var responseEditor = await explorationEditorMainTab.getResponseEditor(0);
+    await responseEditor.markAsCorrect();
+    responseEditor = await explorationEditorMainTab.getResponseEditor(
+      'default');
+    await responseEditor.setFeedback(
+      await forms.toRichText('try again'));
+    responseEditor = await explorationEditorMainTab.getResponseEditor(
+      'default');
+    responseEditor = await explorationEditorMainTab.getResponseEditor(
+      'default');
+    await responseEditor.setDestination(null, false, null);
+
+    await explorationEditorMainTab.moveToState('final card');
+    await explorationEditorMainTab.setInteraction('EndExploration');
+    await explorationEditorPage.saveChanges();
+    await explorationEditorMainTab.moveToState('Introduction');
+
+    var previewTabButton = $('.e2e-test-select-preview-tab-button');
```

Comment:
You shouldn't be referring directly to components in this test. This test should just show the overall flow and the infrastructural parts should be in the corresponding page object (e.g. explorationEditorMainTab).

Please fix here and elsewhere.

Reply:
Thanks @seanlip,
Just to clarify, should I create utility functions in utility files (eg. explorationEditorPage, explorationEditorMainTab, etc) instead of directly referring to the elements and then using those utility function to test the flow in the test (that is coreEditorAndPlayerFeature).
Is this right?

Reply:
I made the required changes. PTAL